nimble/host/src/ble_gattc.c

```diff
@@ -4212,6 +4223,131 @@ ble_gatts_notify_custom(uint16_t conn_handle, uint16_t chr_val_handle,
+            if (rc != 0) {
+                /* Fatal error; application disallowed attribute read. */
+                rc = BLE_HS_EAPP;
+                goto done;
+            }
+        }
+    }
+
+    mtu = ble_att_mtu(conn_handle);
+    pdu_size = sizeof(uint8_t); /* Opcode */
+    split_at = 0;
+
+    for (i = 0; i < num_tuples; i++) {
+        pdu_size += (2 * sizeof(uint16_t)) + OS_MBUF_PKTLEN(tuples[i].value);
+        if (pdu_size > mtu) {
+            /* The notification will need to be split */
+            if (i == split_at) {
+                /* Single notification too large for server,
+                 * cannot send notify without truncating.
+                 */
+                rc = BLE_HS_ENOMEM;
+                goto done;
+            }
+            split_at = i;
+            /* Reinitialize loop */
+            i--;
+            pdu_size = sizeof(uint8_t);
+        }
+    }
+
+    pdu_size = sizeof(uint8_t);
+    split_at = 0;
+    om = ble_hs_mbuf_att_pkt();
+    if(om == NULL) {
+        rc = BLE_HS_ENOMEM;
+        goto done;
+    }
+    for (i = 0; i < num_tuples; i++) {
+        pdu_size += (2 * sizeof(uint16_t)) + OS_MBUF_PKTLEN(tuples[i].value);
+        if (pdu_size > mtu) {
+            rc = ble_att_clt_tx_multi_notify(conn_handle, om);
+            if (rc != 0) {
+                goto done;
+            }
+            split_at = i;
+            i--;
+            pdu_size = sizeof(uint8_t);
+            om = ble_hs_mbuf_att_pkt();
+            if(om == NULL) {
+                rc = BLE_HS_ENOMEM;
+                goto done;
+            }
+            continue;
+        }
+        /* Handle */
+        rc = os_mbuf_copyinto(om, OS_MBUF_PKTLEN(om),
+                              &tuples[i].handle, sizeof tuples[i].handle);
+        if (rc != 0) {
+            rc = BLE_HS_ENOMEM;
+            os_mbuf_free_chain(om);
+            goto done;
+        }
+        /* Length */
+        rc = os_mbuf_copyinto(om, OS_MBUF_PKTLEN(om),
+                              &(OS_MBUF_PKTLEN(tuples[i].value)),
+                              sizeof(OS_MBUF_PKTLEN(tuples[i].value)));
+        if (rc != 0) {
+            rc = BLE_HS_ENOMEM;
+            os_mbuf_free_chain(om);
+            goto done;
+        }
+        /* Value */
+        rc = os_mbuf_appendfrom(om, tuples[i].value,
+                                0, OS_MBUF_PKTLEN(tuples[i].value));
+        if (rc != 0) {
+            rc = BLE_HS_ENOMEM;
+            os_mbuf_free_chain(om);
+            goto done;
+        }
+    }
+    rc = ble_att_clt_tx_multi_notify(conn_handle, om);
+
+done:
+    if (rc != 0) {
+        STATS_INC(ble_gattc_stats, multi_notify_fail);
+    }
+
+    /* Tell the application that multiple notification transmissions were attempted. */
+    for (i = 0; i < num_tuples; i++) {
+        ble_gap_notify_tx_event(rc, conn_handle, tuples[i].handle, 0);
+        os_mbuf_free_chain(tuples[i].value);
+    }
```

Comment:
```suggestion
    int j; //this on top
    /* Tell the application that multiple notification transmissions were attempted. */
    for (j = 0; j < i; j++) {
        ble_gap_notify_tx_event(rc, conn_handle, tuples[j].handle, 0);
        os_mbuf_free_chain(tuples[j].value);
    }
```
`i` now holds how many tuples were actually send (it may be leftover).

Reply:
This is fine because we are assigning 0 to i again